actors/eam/src/lib.rs

```diff
@@ -0,0 +1,294 @@
+            code.len()
+        )))
+    } else if code.is_empty() {
+        Err(ActorError::illegal_argument("Supplied EVM initcode cannot be empty.".into()))
+    } else {
+        Ok(())
+    }
+}
+
+/// hash of data with Keccack256, with first 12 bytes cropped
+fn hash_20<BS, RT>(rt: &RT, data: &[u8]) -> [u8; 20]
+where
+    BS: Blockstore + Clone,
+    RT: Runtime<BS>,
+{
+    rt.hash(SupportedHashes::Keccak256, data)[12..32].try_into().unwrap()
+}
+
+fn create_actor<BS, RT>(
+    rt: &mut RT,
+    creator: EthAddress,
+    new_addr: EthAddress,
+    initcode: Vec<u8>,
+) -> Result<RawBytes, ActorError>
+where
+    BS: Blockstore + Clone,
+    RT: Runtime<BS>,
+{
+    let constructor_params =
+        RawBytes::serialize(EvmConstructorParams { creator, initcode: initcode.into() })?;
+
+    let init_params = Exec4Params {
+        code_cid: rt.get_code_cid_for_type(Type::EVM),
+        constructor_params,
+        subaddress: new_addr.0.to_vec().into(),
+    };
+
+    let ret: ext::init::Exec4Return = rt
+        .send(
+            &INIT_ACTOR_ADDR,
+            ext::init::EXEC4_METHOD,
+            RawBytes::serialize(&init_params)?,
+            rt.message().value_received(),
+        )?
+        .deserialize()?;
+
+    let ret = RawBytes::serialize(EamReturn::from_exec4(ret, new_addr))?;
+
+    Ok(ret)
+}
+
+/// lookup caller's raw ETH address
+fn get_caller_address<BS, RT>(rt: &RT) -> Result<EthAddress, ActorError>
+where
+    BS: Blockstore + Clone,
+    RT: Runtime<BS>,
+{
+    let caller_id = rt.message().caller().id().unwrap();
+
+    let addr = rt.lookup_address(caller_id);
+
+    match addr.map(|a| *a.payload()) {
+        Some(Payload::Delegated(eth)) => {
+            if eth.namespace() == EAM_ACTOR_ID {
+                Ok(EthAddress(eth.subaddress().try_into().unwrap()))
+            } else {
+                Err(ActorError::assertion_failed("Caller is not in the EVM namespace.".into()))
+            }
+        }
+        _ => Err(ActorError::assertion_failed(
+            "All FEVM actors should have a delegated address.".to_string(),
+        )),
+    }
+}
+
+pub struct EamActor;
+impl EamActor {
+    pub fn constructor<BS, RT>(rt: &mut RT) -> Result<(), ActorError>
+    where
+        BS: Blockstore + Clone,
+        RT: Runtime<BS>,
+    {
+        let actor_id = rt.resolve_address(&rt.message().receiver()).unwrap();
+        if actor_id != EAM_ACTOR_ID {
+            return Err(ActorError::forbidden(format!(
+                "The Ethereum Address Manager must be deployed at {EAM_ACTOR_ID}, was deployed at {actor_id}"
+            )));
+        }
+        rt.validate_immediate_caller_type(std::iter::once(&Type::Init))
+    }
+
+    pub fn create<BS, RT>(rt: &mut RT, params: CreateParams) -> Result<RawBytes, ActorError>
+    where
+        BS: Blockstore + Clone,
+        RT: Runtime<BS>,
+    {
+        rt.validate_immediate_caller_type(iter::once(&Type::EVM))?;
+        assert_code_size(&params.initcode)?;
```

Comment:
You know... honestly, I think this should be handled in the EVM actor itself.

Reply:
I.e., the EVM constructor.

Reply:
Yeah this ends up being redundant, since we know that we are calling back into the EVM actor that checks initcode size.

Reply:
though, it does waste a bit more gas on failed calls...